res/res_pjsip_session/pjsip_session_reason_header.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2023, Sangoma Technologies Corporation
+ *
+ * George Joseph <gjoseph@sangoma.com>
+ *
+ * See http://www.asterisk.org for more information about
+ * the Asterisk project. Please do not directly contact
+ * any of the maintainers of this project for assistance;
+ * the project provides a web site, mailing lists and IRC
+ * channels for your use.
+ *
+ * This program is free software, distributed under the terms of
+ * the GNU General Public License Version 2. See the LICENSE file
+ * at the top of the source tree.
+ */
+
+#include "asterisk.h"
+#include "asterisk/res_pjsip_session.h"
+#include "asterisk/utils.h"
+#include "pjsip_session.h"
+
+static const pj_str_t reason_hdr_str = { "Reason", 6};
+
+struct return_reason_data {
+	char *protocol;
+	int response_code;
+	char *response_str;
+	int already_sent;
+};
+
+static void return_reason_destructor(void *obj)
+{
+	struct return_reason_data *rr = obj;
+	SCOPE_ENTER(3, "Destroying RR");
+	ast_free(rr->protocol);
+	ast_free(rr->response_str);
+	ast_free(rr);
+	SCOPE_EXIT("Done");
+}
+
+#define RETURN_REASON_DATASTORE_NAME "pjsip_session_return_reason"
+static struct ast_datastore_info return_reason_info = {
+	.type = RETURN_REASON_DATASTORE_NAME,
+	.destroy = return_reason_destructor,
+};
+
+static void reason_header_outgoing_response(struct ast_sip_session *session,
+	struct pjsip_tx_data *tdata)
+{
+	RAII_VAR(struct ast_datastore *, datastore, NULL, ao2_cleanup);
+	pjsip_generic_string_hdr *reason_hdr;
+	pj_str_t reason_val;
+	RAII_VAR(char *, reason_str, NULL, ast_free);
+	struct return_reason_data *rr = NULL;
+	int rc = 0;
+	struct pjsip_status_line status = tdata->msg->line.status;
+	const char *tag = ast_sip_session_get_name(session);
+	SCOPE_ENTER(3, "%s: Response Code: %d\n", tag,
+		status.code);
+
+	/*
+	 * Include the Reason header if this is a provisional
+	 * response other than a 100 OR it's a 200.
+	 */
+	if (!((PJSIP_IS_STATUS_IN_CLASS(status.code, 100) && status.code != 100) || status.code == 200)) {
+		SCOPE_EXIT_RTN("%s: RC %d not eligible for Reason header\n", tag, status.code);
+	}
+
+	datastore = ast_sip_session_get_datastore(session, RETURN_REASON_DATASTORE_NAME);
+	if (!datastore) {
+		SCOPE_EXIT_RTN("%s: No datastore on session.  Nothing to do\n", tag);
+	}
+	rr = datastore->data;
+
+	rc = ast_asprintf(&reason_str, "%s ;cause=%d ;text=\"%s\"",
```

Comment:
This is borderline offensive.

```
- rc = ast_asprintf(&reason_str, "%s ;cause=%d ;text=\"%s\"",
+ rc = ast_asprintf(&reason_str, "%s; cause=%d; text=\"%s\"",
```

Reply:
That was copied from the RFC.  I unoffensed it though.